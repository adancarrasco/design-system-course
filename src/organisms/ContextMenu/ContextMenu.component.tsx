import * as React from 'react'
import styled from '@emotion/styled'
import { CloseOutlined } from '@material-ui/icons'

import Button, { IButtonProps } from '../../atoms/Button'
import Typography from '../../atoms/Typography'
import Portal from '../../utils/Portal'

export interface IContextMenuProps
  extends IContextMenuStyledProps,
    IMenuContainerStyledProps,
    IBodyContainerStyledProps {
  title: string
  Icon?: React.ReactType
  children: React.ReactNode
  onOpenMenuClick: () => void
  closeButtonColor?: string
  closeButtonColorHover?: string
  buttonToOpenMenu?: React.ReactNode
  buttonProps?: IButtonProps
}

const ContextMenu: React.FC<IContextMenuProps> = ({
  title,
  Icon,
  children,
  onOpenMenuClick,
  ctxMenuWidth,
  enableGutters,
  ctxMenuBackgroundColor,
  closeButtonColor = '#a8a8a8',
  closeButtonColorHover = '#555',
  buttonToOpenMenu,
  buttonProps,
  ...rest
}) => {
  const [isOpen, setIsOpen] = React.useState<boolean>(false)
  const [ctxMenuCoordinates, setCtxMenuCoordinates] = React.useState(null)
  const closeOverlay = () => {
    setIsOpen(false)
    setCtxMenuCoordinates(null)
  }
  const onOpenMenuClickHandler = (e: any) => {
    const targetProps = e.target.getClientRects()[0]
    const { x, y, height } = targetProps
    setCtxMenuCoordinates({ x, y: y + height })
    setIsOpen(true)
    onOpenMenuClick()
  }
  return (
    <ContextMenuStyled {...rest}>
      <OpenMenuButtonStyled {...buttonProps} onClick={onOpenMenuClickHandler} StartIcon={Icon}>
        {buttonToOpenMenu}
      </OpenMenuButtonStyled>
      {isOpen && (
        <Portal>
          <OverlayStyled onClick={closeOverlay} />
          <MenuContainerStyled
            ctxMenuWidth={ctxMenuWidth}
            ctxMenuBackgroundColor={ctxMenuBackgroundColor}
            style={{ left: ctxMenuCoordinates.x, top: ctxMenuCoordinates.y }}
          >
            {title && (
              <TitleContainerStyled>
                <Typography>{title}</Typography>
                <Button
                  backgroundColorHover="transparent"
                  backgroundColor="transparent"
                  color={closeButtonColor}
                  colorHover={closeButtonColorHover}
                  StartIcon={CloseOutlined}
                  onClick={closeOverlay}
                />
              </TitleContainerStyled>
            )}
            <BodyContainerStyled enableGutters={enableGutters}>
              <>{children}</>
            </BodyContainerStyled>
          </MenuContainerStyled>
        </Portal>
      )}
    </ContextMenuStyled>
  )
}

export interface IContextMenuStyledProps {
  backgroundColor?: string
}
const ContextMenuStyled = styled.div<IContextMenuStyledProps>`
  padding: 0.5rem;
  position: relative;
`

const OpenMenuButtonStyled = styled(Button)``

const OverlayStyled = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
`

interface IMenuContainerStyledProps {
  ctxMenuWidth?: string
  ctxMenuBackgroundColor?: string
}
const MenuContainerStyled = styled.div<IMenuContainerStyledProps>`
  position: absolute;
  width: ${(props) => props.ctxMenuWidth};
  background-color: ${(props) => props.ctxMenuBackgroundColor ?? '#fff'};
  box-shadow: 0 8px 16px -4px rgba(9, 30, 66, 0.25), 0 0 0 1px rgba(9, 30, 66, 0.08);
  border-radius: 4px;
`

const TitleContainerStyled = styled.div`
  width: 100%;
  padding: 0.5rem 0;
  margin-bottom: 0.2rem;
  > span {
    display: block;
    box-sizing: border-box;
    margin: 0 0.5rem;
    border-bottom: 1px solid rgba(9, 30, 66, 0.13);
    padding-bottom: 0.6rem;
    text-align: center;
  }
  > button {
    position: absolute;
    right: 0;
    top: 0;
  }
`

interface IBodyContainerStyledProps {
  enableGutters?: boolean
}

const BodyContainerStyled = styled.div<IBodyContainerStyledProps>`
  width: 100%;
  padding-bottom: 0.7rem;
  padding: ${(props) => (props.enableGutters ? '0 0.5rem 0.7rem' : '')};
`

export default ContextMenu
