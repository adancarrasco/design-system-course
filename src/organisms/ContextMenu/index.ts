export { default } from './ContextMenu.component'
export type { IContextMenuProps } from './ContextMenu.component'
