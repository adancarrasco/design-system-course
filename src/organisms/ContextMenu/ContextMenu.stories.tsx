import React from 'react'
import { storiesOf } from '@storybook/react'
import { MoreHoriz } from '@material-ui/icons'

import ContextMenu from '.'
import Button from '../../atoms/Button'
import Badge from '../../atoms/Badge'

storiesOf('Organisms|ContextMenu', module)
  .add('Default', () => (
    <div>
      <ContextMenu
        title="Options"
        Icon={MoreHoriz}
        onOpenMenuClick={() => console.log('Open menu')}
        ctxMenuWidth="300px"
      >
        <Button>Hello</Button>
      </ContextMenu>
    </div>
  ))
  .add('With badge', () => (
    <div>
      <ContextMenu
        title="Options"
        onOpenMenuClick={() => console.log('Open menu')}
        ctxMenuWidth="300px"
        buttonToOpenMenu={<Badge userName="Adan Carrasco" />}
        buttonProps={{
          backgroundColor: 'transparent',
          colorHover: 'initial',
          backgroundColorHover: 'transparent',
        }}
      >
        <Button>Hello</Button>
      </ContextMenu>
    </div>
  ))
