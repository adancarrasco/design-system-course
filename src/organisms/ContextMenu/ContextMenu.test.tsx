import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { matchers } from 'jest-emotion'
import '@testing-library/jest-dom/extend-expect'

import ContextMenu, { IContextMenuProps } from '.'
import { MoreHoriz } from '@material-ui/icons'

expect.extend(matchers)

describe('ContextMenu Component', () => {
  const renderComponent = (props: IContextMenuProps, dataTestId?: string) =>
    render(<ContextMenu data-testid={dataTestId} {...props} />)

  let props
  beforeEach(() => {
    props = {
      title: 'Options',
      Icon: MoreHoriz,
      onOpenMenuClick: jest.fn(),
      ctxMenuWidth: '300px',
    }
  })

  it('matches snapshot and default render', () => {
    const dataTestId = 'context-menu'
    const { getByTestId } = renderComponent(props, dataTestId)
    const container = getByTestId('context-menu')
    expect(container).toMatchSnapshot()
    expect(container).toBeInTheDocument()
  })

  it('should open and dismiss', () => {
    window.Element.prototype.getClientRects = function getClientRects() {
      return ([
        {
          bottom: 0,
          height: 0,
          left: 0,
          right: 0,
          top: 0,
          width: 0,
          x: 0,
          y: 0,
          item: null,
        },
      ] as unknown) as DOMRectList
    }
    props.onClick = jest.fn()
    const { getByRole, getByText, getAllByRole } = renderComponent(props)
    const ctxMenuButton = getByRole(/button/)
    fireEvent.click(ctxMenuButton)
    const contextMenu = getByText('Options')
    expect(contextMenu).toBeInTheDocument()
    const [ctxMenuButton2, closeButton] = getAllByRole(/button/)
    fireEvent.click(closeButton)
    expect(ctxMenuButton2.parentElement.parentElement.nextSibling).toBeNull()
  })
})
