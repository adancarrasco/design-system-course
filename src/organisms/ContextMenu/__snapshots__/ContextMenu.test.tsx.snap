// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu Component matches snapshot and default render 1`] = `
.emotion-1 {
  padding: 0.5rem;
  position: relative;
}

.emotion-0 {
  color: #fff;
  background-color: rgba(130,130,130,0.5);
  font-size: 1rem;
  padding: 0.5rem;
  cursor: pointer;
  border: none;
  border-radius: 4px;
}

.emotion-0:hover {
  color: #ababab;
  background-color: rgba(145,145,145,0.5);
}

.emotion-0 svg:first-of-type {
  margin-right: 4px;
}

.emotion-0 svg:last-child {
  margin-left: 4px;
}

.emotion-0 svg:only-child {
  margin: 0;
}

.emotion-0 > svg,
.emotion-0 > span {
  vertical-align: middle;
}

<div
  class="emotion-1"
  data-testid="context-menu"
>
  <button
    class="emotion-0"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </button>
</div>
`;
